### PR TITLE
Introduce deprecation policy to developer guide.

### LINKED ISSUE #10159: libcudf deprecation policy
As discussed in https://github.com/rapidsai/cudf/pull/10152#issuecomment-1023862782, it would be good to introduce a formal deprecation policy to libcudf. Depending on what types of stability guarantees we provide, such a policy could cover:
- Whether code deprecation before removal is required or only done for critical APIs.
- How long before deprecated code should be removed.
- How we advertise deprecations to users (using the `[[deprecated]]` attribute, doxygen notices, etc).

### PR DESCRIPTION
**Pull request by vyasr**
This PR resolves #10159.

### CODE REVIEW

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [inline cpp/docs/DEVELOPER_GUIDE.md:756]: I don't know if we need to document this and can remove if we think it's extraneous. I figured it would be easier to have it here to start in case it merits discussion.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [inline cpp/docs/DEVELOPER_GUIDE.md:753]: This technically should go in DOCUMENTATION.md, not here, but I figured it was a single line and would be more discoverable here. I can add something to DOCUMENTATION.md in addition (or instead) and link here if reviewers think there would be a benefit.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vuule) [inline cpp/docs/DEVELOPER_GUIDE.md:750]: Maybe there's more to say about this. One kind of breaking change is when we remove an API, and this text explains well what should be done in that case. More often, we replace an API, and it's not clear from this text at which point the new API should be added.
Also might be good to mention that the new API should be referenced in the deprecation message.

Comment (vyasr) [reply cpp/docs/DEVELOPER_GUIDE.md]: I've rewritten this to add some more information. Let me know what you think.

**Review by vuule — APPROVED** [review]
👍 for the PR. 
The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion (unless it happened and I missed it).

**Comment by vyasr** [comment]
> +1 for the PR. The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion (unless it happened and I missed it).

Sounds good. I'm happy to get more opinions on this, perhaps in the meeting next week? I do think that promising users that their code won't break without warning is a large part of what makes a deprecation policy valuable, but I definitely don't mean to impose such a policy unilaterally.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (bdice) [reply cpp/docs/DEVELOPER_GUIDE.md:760]: I think the "breaking" policy is valuable to include. It is helpful to know and would be unclear without this note.

**Comment by bdice** [comment]
> The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion

It seems to me that libcudf's willingness to break APIs quickly (and fix them internally for our Python/Java APIs) is important for the library's development, and changing this policy could impose significant maintenance costs that we currently do not have to pay. While I believe deprecations could be important for external C++ consumers (excluding Python/Java), I have not seen explicit external requests for us to offer more warning before breaking libcudf APIs. This is a key point that I think deserves input from @jrhemstad.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (jrhemstad) [inline cpp/docs/DEVELOPER_GUIDE.md:759]: This is where the whole "breaking" vs. "non-breaking" thing starts to "break down" :) 

Using a `[[deprecated]]` keyword on an API will cause a compiler warning by default. If a user has warnings as errors enabled, their build is broken until they disable that warning or update to the new API. Was that a breaking change?

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I'm going to suggest a radically more laissez-faire policy.

> libcudf is constantly growing and improving to improve performance and better meet our users' needs. As a result, we occasionally need to break or entirely remove APIs to respond to new and improved understanding of the functionality we provide. Remaining free to do this is essential to making libcudf an agile library that can rapidly accommodate our users needs. 

> In order to advertise these breaking changes, any PR that breaks or removes an existing API should be tagged as "breaking". This will ensure that the "Breaking" section of the release notes will include a description of what has broken from the past release. 

> On a best effort basis, the libcudf team will notify users of changes that we expect to have significant or widespread effects.

**Comment by vyasr** [comment]
@jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

**Comment by jrhemstad** [comment]
> @jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

Couldn't agree more. It's important to write down what we do even if what we write down is "We reserve the right to break your code at any time without warning", which isn't that far from the truth :)

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [inline cpp/docs/DEVELOPER_GUIDE.md:757]: Please adjust your word wrapping -- suggest keeping it to the same as our .cpp files -- 100 characters.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [reply cpp/docs/DEVELOPER_GUIDE.md]: I think it should go in both. Please also add an example of using the doxygen `deprecated` command to DOCUMENTATION.md

**Comment by harrism** [comment]
> @jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

Jake's proposed policy is basically what we follow today. I am fine with keeping that agile approach, but documenting it. libcudf has the advantage that most users interface with it via Python or Java. As long as we ensure we don't break those users (by adapting / updating bindings in the same release that we change an API), changing the C++ API is usually OK.

There are also other RAPIDS libraries that depend on libcudf, so we should be careful to not break those as well.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/docs/DEVELOPER_GUIDE.md]: Done. I also added it to an API locally and built the docs to verify that it worked as expected. Conveniently, in addition to showing up on the documentation for a particular API it looks like doxygen will generate a page containing all deprecations:

image


image

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/docs/DEVELOPER_GUIDE.md]: I chose to ignore this issue in what I wrote up. Feel free to suggest an alternative if you have a suggestion on what might be better to write here. I think using `[[deprecated]]` is still better than not, since at least for people not compiling with `Wall` we'll be offering a gentler off ramp. For people who are, the situation is no different than an outright removal.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/docs/DEVELOPER_GUIDE.md]: Done.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [inline cpp/docs/DEVELOPER_GUIDE.md:768]:
```suggestion
Advertise breaking changes by labeling any pull request that breaks or removes an existing API with the "breaking" tag. This ensures that the "Breaking" section of the release notes includes
a description of what has broken from the past release. Label pull requests that contain deprecations with the "non-breaking" tag.
```

File: cpp/docs/DOCUMENTATION.md
Comment (harrism) [inline cpp/docs/DOCUMENTATION.md:375]:
```suggestion
to comment blocks for APIs that will be removed in future releases. Mention alternative / replacement APIs in the deprecation comment.
```

**Review by harrism — CHANGES_REQUESTED** [review]

**Review by jrhemstad — APPROVED** [review]

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [inline cpp/docs/DEVELOPER_GUIDE.md]: Line length. I know this is annoying and manual, but the rest of the doc is wrapped at the same column. Wish clang-tidy would do markdown files. This doesn't mention anything about line length: https://github.com/notslang/tidy-markdown

**Comment by harrism** [comment]
@JohnZed would like to review before merging, but has not had time. Let's wait for him.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (vyasr) [reply cpp/docs/DEVELOPER_GUIDE.md]: I think you looked at the specific commit where I pushed your changes and missed the commit afterwards where reformatted all line lengths. Can you take a look at the current version and make sure the line lengths look like you'd expect?

**Comment by vyasr** [comment]
rerun tests

**Review by ttnghia — APPROVED** [review]

**Comment by vyasr** [comment]
This PR is waiting for @JohnZed to take a look after GTC is completed.

**Review by harrism — APPROVED** [review]

**Comment by ttnghia** [comment]
What?

**Comment by JohnZed** [comment]
I think the text around best effort deprecation is appropriate for libcudf. It has far fewer direct callers than, say, the Python API, so we can afford to evolve it more quickly. I would like to see a more formal deprecation policy for the Python side that does ensure we do a 1-2 release deprecation plan. That should not need to block this PR. Within Python, it is often easier to provide a deprecation path, and we have far more callers 

I think the suggestion to live at head is somewhat unnecessary from this deprecation policy and likely overly prescriptive for end users. The linked video is an hour and a half, and it would be difficult if we imply users need to watch a 90 minute video to understand our deprecation approach. I don't think we're ready to meet all of the criteria that e.g. Abseil has in place for live at head. E.g. they commit to shipping refactoring tools to automate upgrades to support API removals and describe a process of "ship a tool, wait for people to apply the tool, then if practical remove the old API." This is much stricter than what is proposed elsewhere in the doc. In the absence of such tools and guarantees of easy upgrades, I suggest just omitting the "live at head" from text and keeping the best effort plan, while really ensuring we make that effort to simplify upgrades when possible.

We can revisit the best effort deprecation in the future if concrete user needs pop up. To a large extent, I think the Python API is the best place for us to have a "buffer" to allow us to upgrade libcudf APIs without impacting users of the most public-facing API, so we are just leaning on that buffer with this policy.

**Comment by vyasr** [comment]
I have removed the recommendation (and link) to live at head. I agree that it's not explicitly necessary.

Regarding Python, I introduced a deprecation policy of 1 release when I wrote the current draft of the Python developer guide, and we have been following that rule (consistently but somewhat unofficially) for the last 3-4 releases. I am currently working on porting that developer guide into a more visible space, at which point it should become less unofficial, but I agree that the deprecation policy is more critical there and so I prioritized getting that in place first.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
